cmd/root.go

```diff
@@ -83,248 +81,190 @@ var formats = [][2]string{
 }
 var tlsVersions = []string{"1.3", "1.2", "1.1", "1.0", "1"}
 
+type CLI struct {
+	config    *config.Config
+	collector *collector.Collector
+	logger    *log.Logger
+
+	promptMode    bool
+	promptHistory []string
+	schemaTree    *yang.Entry
+
+	wg        *sync.WaitGroup
+	printLock *sync.Mutex
+}
+
+var cli = &CLI{
+	config:        config.New(),
+	logger:        log.New(ioutil.Discard, "", log.LstdFlags),
+	promptHistory: make([]string, 0, 128),
+	schemaTree: &yang.Entry{
+		Dir: make(map[string]*yang.Entry),
+	},
+
+	wg:        new(sync.WaitGroup),
+	printLock: new(sync.Mutex),
+}
+
 var cfgFile string
-var f io.WriteCloser
-var logger *log.Logger
-var coll *collector.Collector
 
 func rootCmdPersistentPreRunE(cmd *cobra.Command, args []string) error {
-	debug := viper.GetBool("debug")
-	if viper.GetString("log-file") != "" {
-		var err error
-		f, err = os.OpenFile(viper.GetString("log-file"), os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+	cli.config.SetLogger()
+	cli.config.SetPersistantFlagsFromFile(rootCmd)
+	cli.config.Globals.Address = sanitizeArrayFlagValue(cli.config.Globals.Address)
+	cli.logger = log.New(ioutil.Discard, "gnmic ", log.LstdFlags|log.Lmicroseconds)
+	if cli.config.Globals.LogFile != "" {
+		f, err := os.OpenFile(cli.config.Globals.LogFile, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
 		if err != nil {
 			return fmt.Errorf("error opening log file: %v", err)
 		}
+		cli.logger.SetOutput(f)
 	} else {
-		if debug {
-			viper.Set("log", true)
+		if cli.config.Globals.Debug {
+			cli.config.Globals.Log = true
 		}
-		switch viper.GetBool("log") {
-		case true:
-			f = os.Stderr
-		case false:
-			f = myWriteCloser{ioutil.Discard}
+		if cli.config.Globals.Log {
+			cli.logger.SetOutput(os.Stderr)
 		}
 	}
-	loggingFlags := log.LstdFlags | log.Lmicroseconds
-	if debug {
-		loggingFlags |= log.Llongfile
+	if cli.config.Globals.Debug {
+		cli.logger.SetFlags(log.LstdFlags | log.Lmicroseconds | log.Llongfile)
 	}
-	logger = log.New(f, "gnmic ", loggingFlags)
-	if debug {
-		grpclog.SetLogger(logger) //lint:ignore SA1019 see https://github.com/karimra/gnmic/issues/59
-		log.Printf("version=%s, commit=%s, date=%s, gitURL=%s, docs=https://gnmic.kmrd.dev", version, commit, date, gitURL)
+
+	if cli.config.Globals.Debug {
+		grpclog.SetLogger(cli.logger) //lint:ignore SA1019 see https://github.com/karimra/gnmic/issues/59
```

Comment:
can we tackle #59 in this branch?

Reply:
not yet, for that the loggerv2 of grpc needs to be implemented.
that change is just a replacement of the global var logger with cli.logger